core/trie_verify.go

```diff
@@ -0,0 +1,353 @@
+	vt := &VerifyTask{
+		diffhash:             diffhash,
+		blockHeader:          header,
+		candidatePeers:       peers,
+		BadPeers:             make(map[string]struct{}),
+		db:                   db,
+		allowUntrustedVerify: allowUntrustedVerify,
+		messageCh:            make(chan VerifyMessage),
+		terminalCh:           make(chan struct{}),
+	}
+	go vt.Start(verifyCh)
+	return vt
+}
+
+func (vt *VerifyTask) Start(verifyCh chan common.Hash) {
+	vt.startAt = time.Now()
+
+	vt.selectPeersToVerify(defaultPeerNumber)
+	resend := time.NewTicker(resendInterval)
+	defer resend.Stop()
+	for {
+		select {
+		case msg := <-vt.messageCh:
+			switch msg.verifyResult.Status {
+			case types.StatusFullVerified:
+				vt.compareRootHashAndWrite(msg, verifyCh)
+			case types.StatusUntrustedVerified:
+				log.Warn("block %s , num= %s is untrusted verified", msg.verifyResult.BlockHash, msg.verifyResult.BlockNumber)
+				if vt.allowUntrustedVerify {
+					vt.compareRootHashAndWrite(msg, verifyCh)
+				}
+			case types.StatusDiffHashMismatch, types.StatusImpossibleFork, types.StatusUnexpectedError:
+				vt.BadPeers[msg.peerId] = struct{}{}
+				log.Debug("peer %s is not available: %s", msg.peerId, msg.verifyResult.Status.Msg)
+			case types.StatusBlockTooNew, types.StatusBlockNewer, types.StatusPossibleFork:
+				log.Info("return msg from peer %s for block %s is %s", msg.peerId, msg.verifyResult.BlockHash, msg.verifyResult.Status.Msg)
+			}
+		case <-resend.C:
+			// if a task has run over 15s, try all the vaild peers to verify.
+			if time.Since(vt.startAt) < tryAllPeersTime {
+				vt.selectPeersToVerify(1)
+			} else {
+				vt.selectPeersToVerify(-1)
+			}
+		case <-vt.terminalCh:
+			return
+		}
+	}
+}
+
+// selectPeersAndVerify func select at most n peers from (candidatePeers-badPeers) randomly and send verify request.
+// when n<0, send to all the peers exclude badPeers.
+func (vt *VerifyTask) selectPeersToVerify(n int) {
+	var validPeers []VerifyPeer
+	candidatePeers := vt.candidatePeers.GetVerifyPeers()
+	for _, p := range candidatePeers {
+		if _, ok := vt.BadPeers[p.ID()]; !ok {
+			validPeers = append(validPeers, p)
+		}
+	}
+	// if
+	if n < 0 || n >= len(validPeers) {
+		for _, p := range validPeers {
+			p.RequestRoot(vt.blockHeader.Number.Uint64(), vt.blockHeader.Hash(), vt.diffhash)
+		}
+		return
+	}
+
+	// if n < len(validPeers), select n peers from validPeers randomly.
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(validPeers), func(i, j int) { validPeers[i], validPeers[j] = validPeers[j], validPeers[i] })
+	for i := 0; i < n; i++ {
+		p := validPeers[i]
+		p.RequestRoot(vt.blockHeader.Number.Uint64(), vt.blockHeader.Hash(), vt.diffhash)
+	}
+}
+
+func (vt *VerifyTask) compareRootHashAndWrite(msg VerifyMessage, verifyCh chan common.Hash) {
+	if msg.verifyResult.Root == vt.blockHeader.Root {
+		blockhash := msg.verifyResult.BlockHash
+		rawdb.MarkTrustBlock(vt.db, blockhash)
```

Comment:
we have done `MarkTrustBlock` twice, I think we can delete here, but let VerifyManager to do it.